File: HackerNewsAndroid/app/src/main/java/com/nurtrino/hackernews/net/AlgoliaApi.kt
package com.nurtrino.hackernews.net

import com.nurtrino.hackernews.model.CommentNode
import com.nurtrino.hackernews.model.Item
import com.nurtrino.hackernews.model.ItemKind
import com.nurtrino.hackernews.model.withDescendantCounts
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonObject
import java.net.URLEncoder

enum class SearchScope(val title: String, val tag: String) {
    STORIES("Stories", "story"),
    COMMENTS("Comments", "comment"),
    ASK("Ask HN", "ask_hn"),
    SHOW("Show HN", "show_hn"),
}

enum class SearchSort(val title: String, val path: String) {
    RELEVANCE("Relevance", "search"),
    DATE("Newest", "search_by_date"),
}

enum class SearchPeriod(val title: String, val windowSeconds: Long?) {
    ALL_TIME("All time", null),
    DAY("Past 24h", 86_400),
    WEEK("Past week", 7 * 86_400),
    MONTH("Past month", 30 * 86_400),
    YEAR("Past year", 365 * 86_400),
}

data class SearchResults(
    val items: List<Item> = emptyList(),
    val page: Int = 0,
    val totalPages: Int = 0,
    val totalHits: Int = 0,
) {
    val hasMore: Boolean get() = page + 1 < totalPages
}

/**
 * Client for HN Search (the Algolia-backed index).
 *
 * Two jobs: full-text search, and pulling an entire comment tree in a single
 * request — which is why threads open fast. Walking the Firebase API would
 * need one request per comment.
 */
object AlgoliaApi {
    private const val BASE = "https://hn.algolia.com/api/v1"

    suspend fun search(
        query: String,
        scope: SearchScope,
        sort: SearchSort,
        period: SearchPeriod,
        page: Int = 0,
        hitsPerPage: Int = 30,
        nowSeconds: Long = System.currentTimeMillis() / 1000,
    ): SearchResults {
        val params = buildString {
            append("query=").append(encode(query))
            append("&tags=").append(encode(scope.tag))
            append("&page=").append(page)
            append("&hitsPerPage=").append(hitsPerPage)
            period.windowSeconds?.let {
                append("&numericFilters=").append(encode("created_at_i>${nowSeconds - it}"))
            }
        }
        return parseSearch(Http.getString("$BASE/${sort.path}?$params"))
    }

    /** Stories or comments submitted by one account, newest first. */
    suspend fun submissions(author: String, scope: SearchScope, page: Int = 0): SearchResults {
        val params = "tags=${encode("author_$author,${scope.tag}")}&page=$page&hitsPerPage=30"
        return parseSearch(Http.getString("$BASE/search_by_date?$params"))
    }

    /** The whole comment tree for a story, already flattened into reading order. */
    suspend fun thread(id: Int): Pair<Item?, List<CommentNode>> {
        val body = Http.getString("$BASE/items/$id")
        val root = Http.json.parseToJsonElement(body) as? JsonObject
            ?: return null to emptyList()

        val flat = mutableListOf<CommentNode>()
        flatten(root["children"] as? JsonArray, 0, flat)
        return rootAsStory(root) to flat.withDescendantCounts()
    }

    private fun flatten(children: JsonArray?, depth: Int, out: MutableList<CommentNode>) {
        if (children == null) return
        for (element in children) {
            val node = element as? JsonObject ?: continue
            val nodeId = node.int("id") ?: continue
            val text = node.str("text").orEmpty()
            val author = node.str("author")
            val kids = node["children"] as? JsonArray
            val isDeleted = author == null && text.isEmpty()

            // A deleted comment with no surviving replies adds nothing; one
            // that still has replies stays as a placeholder so the thread
            // keeps its shape.
            if (isDeleted && (kids == null || kids.isEmpty())) continue

            out.add(
                CommentNode(
                    id = nodeId,
                    parent = node.int("parent_id"),
                    author = author,
                    html = text,
                    time = node.long("created_at_i"),
                    depth = depth,
                    isDeleted = isDeleted,
                )
            )
            flatten(kids, depth + 1, out)
        }
    }

    /**
     * The root of a thread response doubles as the story record, which lets a
     * deep link open a discussion without a second round trip.
     */
    private fun rootAsStory(root: JsonObject): Item? {
        val id = root.int("id") ?: return null
        val type = root.str("type")
        if (type == "comment") return null
        return Item(
            id = id,
            kind = ItemKind.parse(type),
            by = root.str("author"),
            time = root.long("created_at_i"),
            text = root.str("text"),
            url = root.str("url"),
            score = root.int("points"),
            title = root.str("title"),
        )
    }

    private fun parseSearch(body: String): SearchResults {
        val obj = runCatching { Http.json.parseToJsonElement(body) as? JsonObject }.getOrNull()
            ?: return SearchResults()
        val hits = (obj["hits"] as? JsonArray).orEmpty()
        return SearchResults(
            items = hits.mapNotNull { hitToItem(it as? JsonObject) },
            page = obj.int("page") ?: 0,
            totalPages = obj.int("nbPages") ?: 0,
            totalHits = obj.int("nbHits") ?: 0,
        )
    }

    private fun hitToItem(hit: JsonObject?): Item? {
        if (hit == null) return null
        val id = hit.str("objectID")?.toIntOrNull() ?: return null
        val tags = hit.strList("_tags")
        val kind = when {
            tags.contains("comment") -> ItemKind.COMMENT
            tags.contains("job") -> ItemKind.JOB
            else -> ItemKind.STORY
        }
        return Item(
            id = id,
            kind = kind,
            by = hit.str("author"),
            time = hit.long("created_at_i"),
            text = hit.str("story_text") ?: hit.str("comment_text"),
            parent = hit.int("parent_id"),
            url = hit.str("url"),
            score = hit.int("points"),
            title = hit.str("title"),
            descendants = hit.int("num_comments"),
            storyTitle = hit.str("story_title"),
            storyId = hit.int("story_id"),
        )
    }

    private fun encode(value: String): String = URLEncoder.encode(value, "UTF-8")
}

File: HackerNews/Networking/AlgoliaAPI.swift
import Foundation

enum SearchScope: String, CaseIterable, Identifiable, Sendable {
    case stories, comments, ask, show

    var id: String { rawValue }

    var title: String {
        switch self {
        case .stories: return "Stories"
        case .comments: return "Comments"
        case .ask: return "Ask HN"
        case .show: return "Show HN"
        }
    }

    var tag: String {
        switch self {
        case .stories: return "story"
        case .comments: return "comment"
        case .ask: return "ask_hn"
        case .show: return "show_hn"
        }
    }
}

enum SearchSort: String, CaseIterable, Identifiable, Sendable {
    case relevance, date

    var id: String { rawValue }
    var title: String { self == .relevance ? "Relevance" : "Newest" }
    var path: String { self == .relevance ? "search" : "search_by_date" }
}

enum SearchPeriod: String, CaseIterable, Identifiable, Sendable {
    case allTime, day, week, month, year

    var id: String { rawValue }

    var title: String {
        switch self {
        case .allTime: return "All time"
        case .day: return "Past 24h"
        case .week: return "Past week"
        case .month: return "Past month"
        case .year: return "Past year"
        }
    }

    /// Seconds of look-back, or nil for no time filter.
    var window: TimeInterval? {
        switch self {
        case .allTime: return nil
        case .day: return 86_400
        case .week: return 7 * 86_400
        case .month: return 30 * 86_400
        case .year: return 365 * 86_400
        }
    }
}

struct SearchResults: Sendable {
    var items: [Item] = []
    var page = 0
    var totalPages = 0
    var totalHits = 0

    var hasMore: Bool { page + 1 < totalPages }
}

/// Client for HN Search (the Algolia-backed index at hn.algolia.com).
///
/// Two jobs: full-text search, and pulling an entire comment tree in a single
/// request. The latter is why threads open fast — walking the Firebase API
/// would need one request per comment.
actor AlgoliaAPI {
    static let shared = AlgoliaAPI()

    private let base = URL(string: "https://hn.algolia.com/api/v1")!
    private let session: URLSession

    init() {
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = 20
        config.timeoutIntervalForResource = 45
        config.waitsForConnectivity = true
        config.urlCache = URLCache(
            memoryCapacity: 8 * 1024 * 1024,
            diskCapacity: 64 * 1024 * 1024,
            diskPath: "hn-search"
        )
        session = URLSession(configuration: config)
    }

    // MARK: - Search

    func search(
        query: String,
        scope: SearchScope,
        sort: SearchSort,
        period: SearchPeriod,
        page: Int = 0,
        hitsPerPage: Int = 30
    ) async throws -> SearchResults {
        var components = URLComponents(
            url: base.appendingPathComponent(sort.path),
            resolvingAgainstBaseURL: false
        )!
        var queryItems = [
            URLQueryItem(name: "query", value: query),
            URLQueryItem(name: "tags", value: scope.tag),
            URLQueryItem(name: "page", value: String(page)),
            URLQueryItem(name: "hitsPerPage", value: String(hitsPerPage)),
        ]
        if let window = period.window {
            let cutoff = Int(Date().addingTimeInterval(-window).timeIntervalSince1970)
            queryItems.append(URLQueryItem(name: "numericFilters", value: "created_at_i>\(cutoff)"))
        }
        components.queryItems = queryItems

        let data = try await get(components.url!)
        let response = try JSONDecoder().decode(SearchResponse.self, from: data)
        return SearchResults(
            items: response.hits.compactMap { $0.asItem },
            page: response.page,
            totalPages: response.nbPages,
            totalHits: response.nbHits
        )
    }

    /// Stories or comments submitted by one account, newest first.
    func submissions(by author: String, scope: SearchScope, page: Int = 0) async throws -> SearchResults {
        var components = URLComponents(
            url: base.appendingPathComponent("search_by_date"),
            resolvingAgainstBaseURL: false
        )!
        components.queryItems = [
            URLQueryItem(name: "tags", value: "author_\(author),\(scope.tag)"),
            URLQueryItem(name: "page", value: String(page)),
            URLQueryItem(name: "hitsPerPage", value: "30"),
        ]
        let data = try await get(components.url!)
        let response = try JSONDecoder().decode(SearchResponse.self, from: data)
        return SearchResults(
            items: response.hits.compactMap { $0.asItem },
            page: response.page,
            totalPages: response.nbPages,
            totalHits: response.nbHits
        )
    }

    // MARK: - Threads

    /// The whole comment tree for a story in one request, already flattened
    /// into reading order.
    func thread(for id: Int) async throws -> (story: Item?, comments: [CommentNode]) {
        let url = base.appendingPathComponent("items/\(id)")
        let data = try await get(url)
        let root = try JSONDecoder().decode(ThreadNode.self, from: data)

        var flat: [CommentNode] = []
        flatten(root.children, depth: 0, into: &flat)
        return (root.asStory, flat.withDescendantCounts())
    }

    private func flatten(_ nodes: [ThreadNode], depth: Int, into out: inout [CommentNode]) {
        for node in nodes {
            let text = node.text ?? ""
            let isDeleted = node.author == nil && text.isEmpty
            // Deleted comments with no surviving replies add nothing; deleted
            // comments that still have replies are kept as placeholders so the
            // thread doesn't lose its shape.
            if isDeleted && node.children.isEmpty { continue }
            out.append(
                CommentNode(
                    id: node.id,
                    parent: node.parentID,
                    author: node.author,
                    html: text,
                    time: node.createdAt.map { Date(timeIntervalSince1970: $0) },
                    depth: depth,
                    isDeleted: isDeleted
                )
            )
            flatten(node.children, depth: depth + 1, into: &out)
        }
    }

    // MARK: - Plumbing

    private func get(_ url: URL) async throws -> Data {
        do {
            let (data, response) = try await session.data(from: url)
            if let http = response as? HTTPURLResponse, !(200..<300).contains(http.statusCode) {
                throw HNError.badStatus(http.statusCode)
            }
            return data
        } catch let error as URLError
            where error.code == .notConnectedToInternet || error.code == .networkConnectionLost {
            throw HNError.offline
        }
    }
}

// MARK: - Wire format

private struct SearchResponse: Decodable {
    var hits: [SearchHit] = []
    var page = 0
    var nbPages = 0
    var nbHits = 0

    private enum CodingKeys: String, CodingKey { case hits, page, nbPages, nbHits }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        hits = c.lenient([SearchHit].self, .hits) ?? []
        page = c.lenient(Int.self, .page) ?? 0
        nbPages = c.lenient(Int.self, .nbPages) ?? 0
        nbHits = c.lenient(Int.self, .nbHits) ?? 0
    }
}

private struct SearchHit: Decodable {
    var objectID: String = ""
    var title: String?
    var url: String?
    var author: String?
    var points: Int?
    var numComments: Int?
    var createdAt: Double?
    var storyText: String?
    var commentText: String?
    var storyTitle: String?
    var storyID: Int?
    var parentID: Int?
    var tags: [String] = []

    private enum CodingKeys: String, CodingKey {
        case objectID, title, url, author, points
        case numComments = "num_comments"
        case createdAt = "created_at_i"
        case storyText = "story_text"
        case commentText = "comment_text"
        case storyTitle = "story_title"
        case storyID = "story_id"
        case parentID = "parent_id"
        case tags = "_tags"
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        objectID = c.lenient(String.self, .objectID) ?? ""
        title = c.lenient(String.self, .title)
        url = c.lenient(String.self, .url)
        author = c.lenient(String.self, .author)
        points = c.lenient(Int.self, .points)
        numComments = c.lenient(Int.self, .numComments)
        createdAt = c.lenient(Double.self, .createdAt)
        storyText = c.lenient(String.self, .storyText)
        commentText = c.lenient(String.self, .commentText)
        storyTitle = c.lenient(String.self, .storyTitle)
        storyID = c.lenient(Int.self, .storyID)
        parentID = c.lenient(Int.self, .parentID)
        tags = c.lenient([String].self, .tags) ?? []
    }

    var asItem: Item? {
        guard let id = Int(objectID) else { return nil }
        let isComment = tags.contains("comment")
        return Item(
            id: id,
            kind: isComment ? .comment : (tags.contains("job") ? .job : .story),
            by: author,
            time: createdAt.map { Date(timeIntervalSince1970: $0) },
            text: storyText ?? commentText,
            parent: parentID,
            url: url,
            score: points,
            title: title,
            descendants: numComments,
            storyTitle: storyTitle,
            storyID: storyID
        )
    }
}

private struct ThreadNode: Decodable {
    var id: Int = 0
    var type: String?
    var author: String?
    var title: String?
    var url: String?
    var text: String?
    var points: Int?
    var parentID: Int?
    var createdAt: Double?
    var children: [ThreadNode] = []

    private enum CodingKeys: String, CodingKey {
        case id, type, author, title, url, text, points, children
        case parentID = "parent_id"
        case createdAt = "created_at_i"
    }

    init(from decoder: Decoder) throws {
        let c = try decoder.container(keyedBy: CodingKeys.self)
        id = c.lenient(Int.self, .id) ?? 0
        type = c.lenient(String.self, .type)
        author = c.lenient(String.self, .author)
        title = c.lenient(String.self, .title)
        url = c.lenient(String.self, .url)
        text = c.lenient(String.self, .text)
        points = c.lenient(Int.self, .points)
        parentID = c.lenient(Int.self, .parentID)
        createdAt = c.lenient(Double.self, .createdAt)
        children = c.lenient([ThreadNode].self, .children) ?? []
    }

    /// The root node doubles as the story record, which lets a deep link open
    /// a thread without a second round trip.
    var asStory: Item? {
        guard id > 0, type != "comment" else { return nil }
        return Item(
            id: id,
            kind: ItemKind.parse(type),
            by: author,
            time: createdAt.map { Date(timeIntervalSince1970: $0) },
            text: text,
            url: url,
            score: points,
            title: title
        )
    }
}
